libr/bin/dwarf.c

```diff
@@ -1318,321 +1486,537 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
+			memset (&curr_die->attr_values[i], 0, sizeof (curr_die->attr_values[i]));
+
+			buf = parse_attr_value (buf, buf_end - buf,
+				&curr_abbr->defs[i],
+				&curr_die->attr_values[i],
+				&unit->hdr, debug_str, debug_str_len);
+
+			bool is_string_form = curr_die->attr_values[i].attr_form == DW_FORM_strp ||
+				curr_die->attr_values[i].attr_form == DW_FORM_string;
+
+			// TODO  does this have a purpose anymore?
+			// Or atleast it needs to rework becase there will be
+			// more comp units -> more comp dirs and only the last one will be kept
+			if (curr_die->attr_values[i].attr_name == DW_AT_comp_dir &&
+				is_string_form &&
+				curr_die->attr_values[i].string.content) {
+				const char *name = curr_die->attr_values[i].string.content;
 				if ((size_t)name > 1024) { // solve some null derefs
-					sdb_set (s, "DW_AT_comp_dir", name, 0);
+					sdb_set (sdb, "DW_AT_comp_dir", name, 0);
 				} else {
 					eprintf ("Invalid string pointer at %p\n", name);
 				}
 			}
-			cu->dies[cu->length].length++;
+			curr_die->length++;
 		}
-		cu->length++;
+		unit->length++;
 	}
 	return buf;
 }
 
-R_API int r_bin_dwarf_parse_info_raw(Sdb *s, RBinDwarfDebugAbbrev *da,
+/**
+ * @brief Reads all information about compilation unit header
+ * 
+ * @param buf Start of the buffer
+ * @param buf_end Upper bound of the buffer
+ * @param unit Unit to read information into
+ * @return ut8* Advanced position in a buffer
+ */
+static const ut8 *info_comp_unit_read_hdr(const ut8 *buf, const ut8 *buf_end, RBinDwarfCompUnitHdr *hdr) {
+	// 32-bit vs 64-bit dwarf formats
+	// http://www.dwarfstd.org/doc/Dwarf3.pdf section 7.4
+	hdr->length = READ32 (buf);
+	if (hdr->length == (ut32)DWARF_INIT_LEN_64) { // then its 64bit
+		hdr->length = READ64 (buf);
+		hdr->is_64bit = true;
+	}
+	const ut8 *tmp = buf; // to calculate header size
+	hdr->version = READ16 (buf);
+	if (hdr->version == 5) {
+		hdr->unit_type = READ8 (buf);
+
+		hdr->address_size = READ8 (buf);
+
+		if (hdr->is_64bit) {
+			hdr->abbrev_offset = READ64 (buf);
+		} else {
+			hdr->abbrev_offset = READ32 (buf);
+		}
+
+		if (hdr->unit_type == DW_UT_skeleton || hdr->unit_type == DW_UT_split_compile) {
+			hdr->dwo_id = READ8 (buf);
+		} else if (hdr->unit_type == DW_UT_type || hdr->unit_type == DW_UT_split_type) {
+			hdr->type_sig = READ64 (buf);
+
+			if (hdr->is_64bit) {
+				hdr->type_offset = READ64 (buf);
+			} else {
+				hdr->type_offset = READ32 (buf);
+			}
+		}
+	} else {
+		if (hdr->is_64bit) {
+			hdr->abbrev_offset = READ64 (buf);
+		} else {
+			hdr->abbrev_offset = READ32 (buf);
+		}
+		hdr->address_size = READ8 (buf);
+	}
+	hdr->header_size = buf - tmp; // header size excluding length field
+	return buf;
+}
+static int expand_info(RBinDwarfDebugInfo *info) {
+	r_return_val_if_fail (info && info->capacity == info->length, EXIT_FAILURE);
+
+	RBinDwarfCompUnit *tmp;
+	tmp = realloc (info->comp_units, info->capacity * 2 * sizeof (RBinDwarfCompUnit));
+
+	if (!tmp) {
+		return -ENOMEM;
```

Comment:
I see `ENOMEM` and in general all the `-E*` values are used a lot in this file. However, we don't use usually use them in r2 codebase. We just return "negative value" or some special value.

Reply:
Yea, they are left overs from the originil code and at each of the places they are used is XX comment if they should be used. I forgot to ask about, I'll refactor it